proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,121 @@
+TP3: xds-error-propagation
+----
+* Author(s): anicr7
+* Approvers: markdroth, adisuissa
+* Implemented in: <xDS client, ...>
+* Last updated: 2024-10-25
+
+## Abstract
+
+This proposal introduces a mechanism for the xDS control plane to communicate errors to clients in cases where the control plane is unable to send a resource that the client has subscribed to (e.g., because the resource does not exist or because the client lacks permission for the resource), thus improving debuggability. In addition, this also provides a mechanism in the [SotW protocol variants](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#variants-of-the-xds-transport-protocol) for the control plane to explicitly tell the client that a resource does not exist, without the client needing to wait for a [15-second does-not-exist timeout](https://www.envoyproxy.io/docs/envoy/latest/api-docs/xds_protocol#knowing-when-a-requested-resource-does-not-exist), similar to what already exists in the incremental protocol variants.
+
+The objective of this proposal is to suggest a way for clients to receive feedback from xDS Management Servers in case of partial/drastic failures without closing any streams or connections.
+
+This proposal includes a new field for each subscribed Resource, called `ResourceError` which will provide detailed information for resource specific issues. The client must use this additional field to obtain notification for resources the xDS Management server couldn’t procure and provide necessary notification to the application. 
+
+## Background
+
+In cases where the xDS control plane is unable to send a resource that the client has subscribed to (e.g., because the resource does not exist or because the client lacks permission for the resource), the control plane does not have a good mechanism for conveying the cause of the error to the client. The only existing mechanism for conveying a specific error message to a client is for the control plane to close the entire xDS stream with a non-OK status. However, this is often undesirable behavior, because a problem with a single resource will also stop the client from receiving updates from other, unrelated resources.
```

Comment:
> I think the server always knows whether the resource exists at a given point in time. It may be that it will soon discover that the resource has come into existance, but that's a change that it should react to when it occurs. Prior to that point, the resource doesn't exist, and the server should react accordingly.

I think that in eventually consistent distributed systems, in cases where the server subscribes to get the resource from a different source-of-truth, then the server may not know if the resource exists or not until some timeout occurs. If the server is sending "resource doesn't exists" too early, then the client will update the resource state accordingly.

> I don't think anything in this proposal actually affects the fallback decision. This proposal provides a better mechanism to determine that a resource doesn't exist, but regardless of what mechanism we use to determine that, an individual resource not existing does not affect the fallback decision. We don't ever want to fall back just because a single resource doesn't exist.

I'm missing to see where a single resource doesn't exist is discussed here, the above example also applies to cases where all resources of some type are temporarily unknown by the (cold) server, and it sends "unknown" for all the resources before it actually learns about them.